common/pci.c

```diff
@@ -0,0 +1,160 @@
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <list.h>
+#include <mm/slab.h>
+#include <pci.h>
+#include <pci_cfg.h>
+#include <string.h>
+
+#define PCI_NR_BUS  256
+#define PCI_NR_DEV  32
+#define PCI_NR_FUNC 8
+
+#define PCI_HDR_NORMAL     0x00
+#define PCI_HDR_PCI_BRIDGE 0x01
+#define PCI_HDR_MASK       0x7F
+#define PCI_HDR_MULTIFUNC  0x80
+
+#define PCI_STATUS_CAP_LIST (1U << 4)
+
+static list_head_t pci_list = LIST_INIT(pci_list);
+
+static inline bool pci_dev_hdr_type(pcidev_t *dev) {
+    return (dev->hdr_type & PCI_HDR_MASK);
+}
+
+static inline bool pci_dev_is_multifunc(pcidev_t *dev) {
+    return (dev->hdr_type & PCI_HDR_MULTIFUNC) != 0;
+}
+
+static void probe_pci_bus(uint8_t bus, uint8_t start_dev, pcidev_t *bridge) {
+    for (uint8_t dev = start_dev; dev < PCI_NR_DEV; dev++) {
+        for (uint8_t func = 0; func < PCI_NR_FUNC; func++) {
+            pcidev_t *new_dev;
+            uint32_t cfg_val;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0);
+            if ((cfg_val & 0xFFFF) == 0xFFFF) {
+                if (0 == func)
+                    break;
+                continue;
+            }
+
+            new_dev = kzalloc(sizeof(*new_dev));
+            BUG_ON(!new_dev);
+
+            new_dev->segment = 0;
+            new_dev->bus = bus;
+            new_dev->dev = dev;
+            new_dev->func = func;
+            new_dev->vendor_id = cfg_val & 0xFFFF;
+            new_dev->device_id = cfg_val >> 16;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0x1);
+            new_dev->command = cfg_val & 0xFFFF;
+            new_dev->status = cfg_val >> 16;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0x2);
+            new_dev->subclass = (cfg_val & 0xFF0000) >> 16;
+            new_dev->class = cfg_val >> 24;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0x3);
+            new_dev->hdr_type = (cfg_val & 0xFF0000) >> 16;
```

Comment:
nit: I would move these and related to a separate function, like `probe_pci_dev()`